### PR TITLE
PE-2317: Upload Folder does not work on iOS nor Android

### PR DESCRIPTION
**Pull request by matibat**
--- Releases ---
iOS release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:ios:06ea33a95a2e0d42ffce07/releases/2re47p6pgc5l8
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/1q9bpnc47gh7o

### CODE REVIEW

File: ardrive_io/lib/src/io_folder.dart [hunk 1]
```diff
@@ -91,10 +91,8 @@ class _FileSystemFolder extends IOFolder {
     final subFolders = content.whereType<IOFolder>();
     final entities = <T>[];
 
-    if (subFolders.isNotEmpty) {
```
Comment (matibat) [inline ardrive_io/lib/src/io_folder.dart:94]: Unnecessary if conditional

File: ardrive_io/lib/src/io_folder.dart [hunk 2]
```diff
@@ -148,13 +146,15 @@ class IOFolderAdapter {
...
-        folderContent: content);
+      name: getBasenameFromPath(selectedDirectoryPath),
+      lastModifiedDate: (await directory.stat()).modified,
+      path: selectedDirectoryPath,
+      folderContent: content,
+    );
 
     await folder.initFolder();
 
+    print('Picked folder: $folder');
```
Comment (matibat) [inline ardrive_io/lib/src/io_folder.dart:157]: TODO: remove

File: ardrive_io/lib/src/mobile/mobile_io.dart [hunk 3]
```diff
@@ -49,7 +49,13 @@ class MobileIO implements ArDriveIO {
     final provider = _fileProviderFactory.fromSource(FileSource.fileSystem)
         as MultiFileProvider;
 
-    return provider.getFolder();
+    final folder = await provider.getFolder();
+    print('Da picked folder: $folder');
+    print('Da path of the picked folder: ${folder.path}');
+    print('Da content of the picked folder: ${await folder.listContent()}');
+    // throw UnimplementedError();
```
Comment (matibat) [inline ardrive_io/lib/src/mobile/mobile_io.dart:56]: TODO: remove

File: ardrive_io/lib/src/utils/path_utils.dart [hunk 4]
```diff
@@ -19,5 +19,7 @@ String getDirname(String entityPath) {
     throw EntityPathException();
   }
 
-  return path.dirname(entityPath);
+  final dirname = path.dirname(entityPath);
+  print('The dirname of $entityPath: $dirname');
```
Comment (matibat) [inline ardrive_io/lib/src/utils/path_utils.dart:23]: TODO: remove

File: lib/blocs/upload/models/upload_file.dart [hunk 5]
```diff
@@ -1,15 +1,26 @@
...
+  const UploadFile({
+    required this.ioFile,
+    required this.parentFolderId,
+    this.relativeTo,
+  });
 
+  final String? relativeTo;
   final IOFile ioFile;
+
+  // This MF is supposed to hold the EID of its parent
```
Comment (matibat) [inline lib/blocs/upload/models/upload_file.dart:13]: wat?

File: lib/blocs/upload/models/upload_file.dart [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
   final String parentFolderId;
 
   String getIdentifier() {
+    print('File path at UploadFile: ${ioFile.path}');
+    print('relativeTo at UploadFile: $relativeTo');
```
Comment (matibat) [inline lib/blocs/upload/models/upload_file.dart:18]: TODO: remove

File: lib/blocs/upload/upload_cubit.dart [hunk 7]
```diff
@@ -186,6 +188,12 @@ class UploadCubit extends Cubit<UploadState> {
       folder.parentFolderId = folders.containsKey(folder.parentFolderPath)
           ? folders[folder.parentFolderPath]!.id
           : _targetFolder.id;
+
+      print(
+        'Parent folder id for path ${folder.parentFolderPath}: ${folder.parentFolderId}',
+      );
+      // throw UnimplementedError();
+
```
Comment (matibat) [inline lib/blocs/upload/upload_cubit.dart:196]: TODO: remove

File: lib/blocs/upload/upload_cubit.dart [hunk 8]
```diff
@@ -217,11 +225,18 @@ class UploadCubit extends Cubit<UploadState> {
...
     final filesToUpload = <UploadFile>[];
     for (var file in files) {
-      final fileFolder = getDirname(file.ioFile.path);
+      final fileFolder = getDirname(file.relativeTo != null
+          ? file.ioFile.path.replaceFirst('${file.relativeTo}/', '')
+          : file.ioFile.path);
+      final parentFolderId = folders[fileFolder]?.id ?? _targetFolder.id;
+      print('The parent folder ID will be: $parentFolderId');
+      print('Path of the parent folder: $fileFolder');
+      print('Folders mapping: $folders');
```
Comment (matibat) [inline lib/blocs/upload/upload_cubit.dart:234]: TODO: remove

File: lib/blocs/upload/upload_cubit.dart [hunk 9]
```diff
@@ -272,6 +287,11 @@ class UploadCubit extends Cubit<UploadState> {
         return;
       }
 
+      print(
+        'There are ${files.length} files and ${foldersByPath.length} folders',
+      );
+      print('Folders: $foldersByPath, files: $files');
+
```
Comment (matibat) [inline lib/blocs/upload/upload_cubit.dart:294]: TODO: remove

File: lib/components/upload_form.dart [hunk 10]
```diff
@@ -20,19 +20,27 @@ import 'components.dart';
...
     final ioFolder = await io.pickFolder();
     final ioFiles = await ioFolder.listFiles();
+    final uploadFiles = ioFiles.map((file) {
+      final uf = UploadFile(
+        ioFile: file,
 
-    final uploadFiles = ioFiles
-        .map((file) => UploadFile(ioFile: file, parentFolderId: folderId))
-        .toList();
+        // We don't yet know the EID of its parent, this field stands for the selected folder on the UI x.x
```
Comment (matibat) [inline lib/components/upload_form.dart:35]: TODO: remove

File: lib/components/upload_form.dart [hunk 11]
```diff
@@ -20,19 +20,27 @@ import 'components.dart';
...
+        parentFolderId: parentFolderId,
 
+        relativeTo: getDirname(ioFolder.path),
+      );
+      print('The dirname of the picked folder: ${getDirname(ioFolder.path)}');
+      print('UploadFile: ${uf.relativeTo}');
```
Comment (matibat) [inline lib/components/upload_form.dart:41]: TODO: remove

File: lib/utils/upload_plan_utils.dart [hunk 12]
```diff
@@ -39,13 +40,19 @@ class UploadPlanUtils {
       // If path is a blob from drag and drop, use file name. Else use the path field from folder upload
       final filePath = '${targetFolder.path}/${file.getIdentifier()}';
 
+      print('Target folder path when planning upload: ${targetFolder.path}');
+      print('File identifier when planning upload: ${file.getIdentifier()}');
+      print('File path when planning upload: $filePath');
+
```
Comment (matibat) [inline lib/utils/upload_plan_utils.dart:46]: TODO: remove

File: lib/utils/upload_plan_utils.dart [hunk 13]
```diff
@@ -82,6 +89,7 @@ class UploadPlanUtils {
       }
     }
     foldersByPath.forEach((key, folder) async {
+      print('Folder to upload: $folder');
```
Comment (matibat) [inline lib/utils/upload_plan_utils.dart:92]: TODO: remove

File: lib/utils/upload_plan_utils.dart [hunk 14]
```diff
@@ -104,12 +112,19 @@ class UploadPlanUtils {
...
     // Generate folders
     for (var file in files) {
-      final path = file.ioFile.path;
+      final relativeTo = file.relativeTo;
+      final path = relativeTo != null
+          ? file.ioFile.path.replaceFirst('$relativeTo/', '')
+          : file.ioFile.path;
+      print('Relative folder path: $path');
+      print('Relative to: $relativeTo');
+      print('File path: ${file.ioFile.path}');
```
Comment (matibat) [inline lib/utils/upload_plan_utils.dart:127]: TODO: remove

**Review by matibat — COMMENTED** [review]
Self review - I still want to test more cases

File: lib/blocs/upload/models/upload_file.dart (hunk 5 above)
Comment (thiagocarvalhodev) [reply lib/blocs/upload/models/upload_file.dart:13]: loved this self review 😂

**Review by matibat — COMMENTED** [review]
I've tested at this point:

It's still working on web; folder uploads on mobile seems to work now.

File: ardrive_io/lib/src/utils/path_utils.dart [hunk 15]
```diff
@@ -18,6 +18,5 @@ String getDirname(String entityPath) {
   if (entityPath.isEmpty) {
     throw EntityPathException();
   }
-
```
Comment (karlprieb) [inline ardrive_io/lib/src/utils/path_utils.dart:21]: removing this empty lines decrease the readability as now we don't have a clear perception of when the previous code block ends.

File: lib/utils/upload_plan_utils.dart [hunk 16]
```diff
@@ -39,13 +40,15 @@ class UploadPlanUtils {
...
+
       final fileSize = await file.ioFile.length;
       final fileEntity = FileEntity(
         driveId: targetDrive.id,
         name: fileName,
         size: fileSize,
         lastModifiedDate: file.ioFile.lastModifiedDate,
-        parentFolderId: file.parentFolderId,
+        parentFolderId: parentFolderId?.id ?? file.parentFolderId,
         dataContentType: lookupMimeType(fileName) ?? 'application/octet-stream',
```
Comment (karlprieb) [inline lib/utils/upload_plan_utils.dart:52]: we can use `lookupMimeTypeWithDefaultType` from `ardrive_io` here :D

File: ardrive_io/lib/src/file_provider.dart [hunk 17]
```diff
@@ -138,7 +141,13 @@ class FilePickerProvider implements MultiFileProvider {
 
     final selectedDirectory = Directory(selectedDirectoryPath);
 
-    return _ioFolderAdapter.fromFileSystemDirectory(selectedDirectory);
+    await SecurityScopedResource.instance
+        .startAccessingSecurityScopedResource(selectedDirectory);
```
Comment (matibat) [inline ardrive_io/lib/src/file_provider.dart:145]: This is already being called in secureScopedAction

Comment (thiagocarvalhodev) [reply ardrive_io/lib/src/file_provider.dart:145]: Good catch!

Comment (thiagocarvalhodev) [reply ardrive_io/lib/src/file_provider.dart]: [722d653](https://github.com/ardriveapp/ardrive-web/pull/697/commits/722d653947455589718c5dcd71f2bb817658b595)

File: ardrive_io/lib/src/utils/path_utils.dart (hunk 15 above)
Comment (matibat) [reply ardrive_io/lib/src/utils/path_utils.dart]: f7f5027b31b32a02d77ff96c1d586883875d729e

File: lib/utils/upload_plan_utils.dart (hunk 16 above)
Comment (matibat) [reply lib/utils/upload_plan_utils.dart]: c3ab2e976af447ed87266e1737d9b7b8e0c99374

**Review by thiagocarvalhodev — APPROVED** [review]

**Review by karlprieb — DISMISSED** [review]

**Review by karlprieb — DISMISSED** [review]

**Review by karlprieb — APPROVED** [review]

**Pull request merged**